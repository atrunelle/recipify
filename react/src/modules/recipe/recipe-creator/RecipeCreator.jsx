import './RecipeCreator.css';

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Grid } from '@material-ui/core';

import SearchIngredient from '../components/search-ingredient/SearchIngredient';
import IngredientsList from '../components/ingredients-list/IngredientsList';
import IngredientsData from '../components/ingredients-data/IngredientsData';
import * as ingredientsActions from '../store/actions/ingredientsActions';
import * as recipesActions from '../store/actions/recipesActions';
import recipeService from '../recipe.service';
import NutritionMacro from '../components/nutrition-macro/NutritionMacro';
import nutritionService from '../nutrition.service';

class RecipeCreator extends Component {
	addIngredient = (ingredientName, numberOfServings) => {
		this.props.dispatch(ingredientsActions.fetchIngredientPending());

		recipeService.fetchIngredient(ingredientName, numberOfServings)
			.then((data) => {
				this.props.dispatch(ingredientsActions.addIngredient(data));
				this.calculateTotalNutriton();
			})
			.catch((error) => {
				this.props.dispatch(ingredientsActions.fetchIngredientRejected(error));
			});
  }

  removeIngredient = (index) => {
		if (this.props.ingredients.length === 1) {
			this.removeAllIngredients();
		} else {
			this.props.dispatch(ingredientsActions.removeIngredient(index));
			this.calculateTotalNutriton();
		}
  }

  removeAllIngredients = () => {
		this.props.dispatch(ingredientsActions.removeAllIngredients());
		this.props.dispatch(ingredientsActions.resetTotalCalories());
		this.props.dispatch(ingredientsActions.resetTotalWeight());
		this.props.dispatch(ingredientsActions.resetTotalNutritients());
	}
	
	saveRecipe = (name) => {
		const recipe = {
			ingredients: [ ...this.props.ingredients ],
			name,
		};
		this.removeAllIngredients();
		this.props.dispatch(recipesActions.saveRecipe(recipe));
	}

	calculateTotalNutriton = () => {
		const totalCalories = nutritionService.getTotalCalories(this.props.ingredients);
    const totalWeight = nutritionService.getTotalWeight(this.props.ingredients);

		this.props.dispatch(ingredientsActions.updateTotalCalories(totalCalories));
		this.props.dispatch(ingredientsActions.updateTotalWeight(totalWeight));

		const totalNutrients = nutritionService.getTotalNutrients(this.props.ingredients, this.props.totalCalories);
		this.props.dispatch(ingredientsActions.updateTotalNutriton(totalNutrients));
	}

	render() {
		return (
			<Grid container spacing={16}>
				<Grid item xs={12}>
					<SearchIngredient
						add={this.addIngredient}
						isFetching={this.props.isFetching}/>
				</Grid>
				<Grid item xs={12}>
					<IngredientsList 
						ingredients={this.props.ingredients}
						remove={this.removeIngredient}
						removeAll={this.removeAllIngredients}
						save={this.saveRecipe}/>
				</Grid>
				<Grid item xs={12}>
					<IngredientsData 
						totalNutrients={this.props.totalNutrients}
						totalWeight={this.props.totalWeight}
						totalCalories={this.props.totalCalories}/>
				</Grid>
				<Grid item xs={12}>
					<NutritionMacro totalNutrients={this.props.totalNutrients}/>
				</Grid>
			</Grid>
		);
	}
}

const mapStateToProps = (state) => ({
	ingredients: state.ingredients.ingredients,
	totalNutrients: state.ingredients.totalNutrients,
	totalWeight: state.ingredients.totalWeight,
	totalCalories: state.ingredients.totalCalories,
	isFetching: state.ingredients.isFetching,
});

export default connect(mapStateToProps)(RecipeCreator);