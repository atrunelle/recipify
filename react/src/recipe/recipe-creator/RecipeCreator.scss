.search {
    padding: 20px;

    &__input {
        height: 40px;
        width: 100%;
    }

    &__label {
        font-weight: bold;
        margin-bottom: 10px;
        display: block;
    }
}

