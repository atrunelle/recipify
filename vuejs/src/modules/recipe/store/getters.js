import * as types from './types';

export default {
  [types.GET_RECIPE]: (state) => state.recipeIngredients,
  [types.GET_RECIPES]: (state) => state.recipes,
};
