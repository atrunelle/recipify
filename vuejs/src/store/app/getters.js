import * as types from './types';

export default {
  [types.GET_ALERT]: (state) => state.alert,
};
