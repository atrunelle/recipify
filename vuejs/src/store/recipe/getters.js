import nutritionService from '@/core/nutrition.service';

export default {
  totalWeight: (state) => {
    return state.ingredients.reduce((sum, ingredient) => {
      return ingredient.nutrients.totalWeight + sum;
    }, 0);
  },

  totalCalories: (state) => {
    return state.ingredients.reduce((sum, ingredient) => {
      return ingredient.nutrients.calories + sum;
    }, 0);
  },

  totalNutrients: (state, getters) => {
    return nutritionService
      .getTotalNutrients(state.ingredients, getters.totalCalories);
  },
};
