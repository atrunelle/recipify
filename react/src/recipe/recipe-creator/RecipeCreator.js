import './RecipeCreator.css';

import React, { Component } from 'react';
import { connect } from 'react-redux';
import apiService from '../../core/api.service';

import SearchIngredient from '../components/search-ingredient/SearchIngredient';
import IngredientsList from '../components/ingredients-list/IngredientsList';
import IngredientsData from '../components/ingredients-data/IngredientsData';
import * as ingredients from '../actions/ingredientsActions';

class RecipeCreator extends Component {

	getIngredients(data) {
		const MEASUREMENT = 'Kilogram';

		const measureUri = data
			.map((d) => d.measures
				.find(m =>  m.label === MEASUREMENT)
			);

		const ingredients = data.map((item, index) => {
			return {
				quantity: 1,
				measureUri: measureUri[index].uri,
				foodURI: item.food.uri,
			}
		});

		return {
			ingredients,			
		}			
	}

	search(value) {
		 apiService.getItem(value).then((data) => {
			const ingredients = this.getIngredients(data.hints);

			apiService.getNutrients(ingredients)
				.then((results) => {
					this.setState({
						value,
						ingredients: results,
					});

				});
		});
  }

  addIngredient = () => {
    this.props.dispatch(ingredients.addIngredient());
  }

  removeIngredient = (index) => {
    this.props.dispatch(ingredients.removeIngredient(index));
  }

  removeAllIngredients = () => {
    this.props.dispatch(ingredients.removeAllIngredients());
  }

	render() {
		return (
			<div>
				<SearchIngredient add={this.addIngredient}/>
				<IngredientsList 
          ingredients={this.props.ingredients}
          remove={this.removeIngredient}
          removeAll={this.removeAllIngredients}/>
        <IngredientsData ingredients={this.props.ingredients}/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	ingredients: state.ingredients,
});

export default connect(mapStateToProps)(RecipeCreator);