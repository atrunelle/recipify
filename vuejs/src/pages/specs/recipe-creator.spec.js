import Vuex from 'vuex';

import { mount, createLocalVue } from '@vue/test-utils';
import getters from '@/modules/recipe/store/getters';

import RecipeCreator from '../recipe-creator';

describe('Component: recipe creator', () => {
  let component;
  let localVue = createLocalVue();
  localVue.use(Vuex);

  const recipeActions = {
    removeIngredient: jest.fn(),
  };

  beforeEach(() => {
    const recipeStore = {
      namespaced: true,
      state: {
        ingredients: [{
          name: 'carrot',
          nutrients: {
            totalNutrients: {
              FAT: {
                quantity: 5,
                unit: 'gr',
                label: 'Fat',
              },
              PROCNT: {
                quantity: 5,
                unit: 'gr',
                label: 'Fat',
              },
            },
            totalDaily: {
              FAT: {
                quantity: 5,
                unit: '%',
                label: 'Fat',
              },
              PROCNT: {
                quantity: 5,
                unit: '%',
                label: 'Fat',
              },
            },
          },
        }],
        totalNutrients: [],
        totalCalories: 0,
        totalWeight: 0,
      },
      getters,
      actions: recipeActions,
    };

    const store = new Vuex.Store({
      modules: {
        recipe: recipeStore,
      },
    });
    component = mount(RecipeCreator, {
      stubs: ['ingredients-list', 'ingredients-data', 'search-ingredient'],
      localVue,
      store,
    });
  });

  it('should match snapshot', () => {
    expect(component.html()).toMatchSnapshot();
  });
});
