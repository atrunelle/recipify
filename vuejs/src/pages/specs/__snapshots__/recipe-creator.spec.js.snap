// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: recipe creator should match snapshot 1`] = `
<div class="layout row wrap">
    <div class="flex xs12">
        <div class="flex xs12">
            <div></div>
        </div>
        <div class="flex xs12 align-center justify-center">
            <div ingredients="[object Object]" remove="function boundFn(a) {
    var l = arguments.length;
    return l
      ? l > 1
        ? fn.apply(ctx, arguments)
        : fn.call(ctx, a)
      : fn.call(ctx)
  }" remove-all="function boundFn(a) {
    var l = arguments.length;
    return l
      ? l > 1
        ? fn.apply(ctx, arguments)
        : fn.call(ctx, a)
      : fn.call(ctx)
  }" save="function boundFn(a) {
    var l = arguments.length;
    return l
      ? l > 1
        ? fn.apply(ctx, arguments)
        : fn.call(ctx, a)
      : fn.call(ctx)
  }"></div>
        </div>
        <!---->
        <!---->
    </div>
</div>
`;
